replication.go

```diff
@@ -0,0 +1,288 @@
+			h := &ServerHeartbeat{ServerWalEnd: uint64(serverWalEnd), ServerTime: uint64(serverTime), ReplyRequested: replyNow}
+			return &ReplicationMessage{ServerHeartbeat: h}, nil
+		}
+	}
+	return
+}
+
+// Wait for a single replication message up to timeout time.
+//
+// Properly using this requires some knowledge of the postgres replication mechanisms,
+// as the client can receive both WAL data (the ultimate payload) and server heartbeat
+// updates. The caller also must send standby status updates in order to keep the connection
+// alive and working.
+//
+// There is also a condition (during startup) which can cause both the replication message
+// to return as nil as well as the error, which is a normal part of the replication protocol
+// startup. It's important the client correctly handle (ignore) this scenario.
+//
+// This returns pgx.ErrNotificationTimeout when there is no replication message by the specified
+// duration.
+func (c *Conn) WaitForReplicationMessage(timeout time.Duration) (r *ReplicationMessage, err error) {
+	var zeroTime time.Time
+
+	deadline := time.Now().Add(timeout)
+
+	// Use SetReadDeadline to implement the timeout. SetReadDeadline will
+	// cause operations to fail with a *net.OpError that has a Timeout()
+	// of true. Because the normal pgx rxMsg path considers any error to
+	// have potentially corrupted the state of the connection, it dies
+	// on any errors. So to avoid timeout errors in rxMsg we set the
+	// deadline and peek into the reader. If a timeout error occurs there
+	// we don't break the pgx connection. If the Peek returns that data
+	// is available then we turn off the read deadline before the rxMsg.
+	err = c.conn.SetReadDeadline(deadline)
+	if err != nil {
+		return nil, err
+	}
+
+	// Wait until there is a byte available before continuing onto the normal msg reading path
+	_, err = c.reader.Peek(1)
+	if err != nil {
+		c.conn.SetReadDeadline(zeroTime) // we can only return one error and we already have one -- so ignore possiple error from SetReadDeadline
+		if err, ok := err.(*net.OpError); ok && err.Timeout() {
+			return nil, ErrNotificationTimeout
+		}
+		return nil, err
+	}
+
+	err = c.conn.SetReadDeadline(zeroTime)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.readReplicationMessage()
+}
+
+// Start a replication connection, sending WAL data to the given replication
+// receiver. The sql string here should be a "START_REPLICATION" command, as
+// per the postgresql docs here:
+// https://www.postgresql.org/docs/9.5/static/protocol-replication.html
+//
+// A typical query would look like:
+// START_REPLICATION SLOT t LOGICAL test_decoder 0/0
+//
+// Once started, the client needs to invoke WaitForReplicationMessage() in order
+// to fetch the WAL and standby status. Also, it is the responsibility of the caller
+// to periodically send StandbyStatus messages to update the replication slot position.
+func (c *Conn) StartReplication(sql string, arguments ...interface{}) (err error) {
```

Comment:
You also need a varadic string arg that would be options to the logical replication plugin. I thought about this, it's probably a good change.